Dsolve constantsimp

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #4838 
Fixes #13691 
Resurrects #13752 
- [x] needs #16008

See also #4435

#### Brief description of what is fixed or changed

Fix odesimp and constant_renumber so that integration constants in the original ODE are not renumbered. On master we have:
```julia
In [1]: C1 = Symbol('C1')                                                                                                                                     

In [2]: dsolve(f(x).diff(x)-C1*f(x))                                                                                                                          
Out[2]: 
           C₂⋅x
f(x) = C₁⋅ℯ 
```
This PR gives
```julia
In [2]: C1 = Symbol('C1')                                                                                                                                     

In [3]: dsolve(f(x).diff(x)-C1*f(x))                                                                                                                          
Out[3]: 
           C₁⋅x
f(x) = C₂⋅ℯ  
```

#### Other comments

I'm not happy with this patch yet. It feels a bit kludgey and I think it can be done in a cleaner way. I started from #13752 and ended up here by the time I had it working (not much from the original diff remains). It took a little while to get it to work though so I'm just putting it here for now to see how it copes on Travis.

This doesn't yet work with any ODE solved by the nth_algebraic solver due to #15999. If #16008 gets merged I'll rebase that in here and clean this up.

Also possibly there are more open issues that this could close.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* solvers
    * dsolve, constantsimp and odesimp no longer renumber inregration constants found in the original ODE.
<!-- END RELEASE NOTES -->

### TODO
- [x] delete exclude code
- [x] rename nonconstants to variables
- [ ] update docstrings
- [x] add tests

saurabhkgp21 committed:
Renumber only newly introduced constants in dsolve

oscarbenjamin committed:
Fix renumber only newly introduced constants

oscarbenjamin commented:
Also fixes #11282

oscarbenjamin committed:
Add tests for constant renumbering in dsolve

oscarbenjamin commented:
I've rebased to include #16008 and added some tests.

XtremeGood commented:
Does this issue needs work?I 'm interested in it,if it needs so?

XtremeGood commented:
Does this have problem in tests or the function code?

XtremeGood commented:
So what does it needs?I mean you are getting the differential equation solve.'C1' is not messing up with 'C2' then what is the problem ?Would you like to create a new class of integration constants?

oscarbenjamin commented:
Right now the problem is that when I combine this with #15992 (by rebasing that PR on top of this one) the new tests there don't pass. Actually I just realised problems with the tests there so I can push a fix to those. That leaves only one case that fails:
```console
$ pytest sympy/solvers/tests/test_ode.py -m 'not slow or not xfail' -n4
============================================================= test session starts =============================================================
platform darwin -- Python 3.7.1, pytest-4.1.2.dev19+g57bf9d67, py-1.7.0, pluggy-0.8.0
architecture: 64-bit
cache:        yes
ground types: gmpy 1.17

rootdir: /Users/enojb/current/sympy/sympy, inifile: pytest.ini
plugins: xdist-1.26.0, forked-0.2, cov-2.6.1, doctestplus-0.3.0.dev0
gw0 [122] / gw1 [122] / gw2 [122] / gw3 [122]
...........x............................xx.....s..............x..........x....x..........................F..x..xx..xx.....              [100%]
================================================================== FAILURES ===================================================================
___________________________________________________________ test_order_reducible_10 ___________________________________________________________
[gw1] darwin -- Python 3.7.1 /Users/enojb/current/sympy/venv/bin/python3

    def test_order_reducible_10():
        eqn = f(x).diff(x, 5) + 2*f(x).diff(x, 3) + f(x).diff(x)
        sol = Eq(f(x), C1 + (C2 + C3*x)*sin(x) + (C4 + C5*x)*cos(x))
        sols = constant_renumber(sol, 'C', 1, 5)
        assert checkodesol(eqn, sol, order=2, solve_for_func=False) == (True, 0)
        assert dsolve(eqn, f(x)) in (sol, sols)
>       assert dsolve(eqn, f(x), hint='order_reducible') in (sol, sols)
E       AssertionError: assert Eq(f(x), C1 + C2*(x*sin(x) + cos(x)) + C3*(-x*cos(x) + sin(x)) + C4*sin(x) + C5*cos(x)) in (Eq(f(x), C1 + (C2 + C3*x)*sin(x) + (C4 + C5*x)*cos(x)), Eq(f(x), C1 + (C2 + C3*x)*sin(x) + (C4 + C5*x)*cos(x)))
E        +  where Eq(f(x), C1 + C2*(x*sin(x) + cos(x)) + C3*(-x*cos(x) + sin(x)) + C4*sin(x) + C5*cos(x)) = dsolve(Derivative(f(x), x) + 2*Derivative(f(x), (x, 3)) + Derivative(f(x), (x, 5)), f(x), hint='order_reducible')
E        +    where f(x) = f(x)

sympy/solvers/tests/test_ode.py:3219: AssertionError
                                                               DO *NOT* COMMIT!                                                               
======================================== 1 failed, 109 passed, 1 skipped, 11 xfailed in 306.81 seconds ========================================
```
So with these two PRs combined order_reducible returns
```python
Eq(f(x), C1 + C2*(x*sin(x) + cos(x)) + C3*(-x*cos(x) + sin(x)) + C4*sin(x) + C5*cos(x))
```
whereas current dsolve returns
```python
Eq(f(x), C1 + (C2 + C3*x)*sin(x) + (C4 + C5*x)*cos(x))
```

We need to decide if these are equivalent general solutions:
```julia
In [2]: C1, C2, C3, C4, C5 = symbols('C1:6')                                                                                                   

In [3]: sol_reducible = Eq(f(x), C1 + C2*(x*sin(x) + cos(x)) + C3*(-x*cos(x) + sin(x)) + C4*sin(x) + C5*cos(x))                                

In [4]: sol_dsolve = Eq(f(x), C1 + (C2 + C3*x)*sin(x) + (C4 + C5*x)*cos(x))                                                                    

In [5]: expand(sol_reducible.rhs)                                                                                                              
Out[5]: C₁ + C₂⋅x⋅sin(x) + C₂⋅cos(x) - C₃⋅x⋅cos(x) + C₃⋅sin(x) + C₄⋅sin(x) + C₅⋅cos(x)

In [6]: expand(sol_dsolve.rhs)                                                                                                                 
Out[6]: C₁ + C₂⋅sin(x) + C₃⋅x⋅sin(x) + C₄⋅cos(x) + C₅⋅x⋅cos(x)
```
`checkodesol` thinks that both are solutions:
```julia
In [7]: eq =  f(x).diff(x, 5) + 2*f(x).diff(x, 3) + f(x).diff(x)                                                                               

In [8]: checkodesol(eq, sol_dsolve)                                                                                                            
Out[8]: (True, 0)

In [9]: checkodesol(eq, sol_reducible)                                                                                                         
Out[9]: (True, 0)
```
but are they equally general?
```julia
C₁ + C₂⋅x⋅sin(x) + C₂⋅cos(x) - C₃⋅x⋅cos(x) + C₃⋅sin(x) + C₄⋅sin(x) + C₅⋅cos(x)
C₁ + C₂⋅sin(x) + C₃⋅x⋅sin(x) + C₄⋅cos(x) + C₅⋅x⋅cos(x)
```
Okay so if we match them as `C1:C1`,  `(C3+C4):C2`, `C2:C3`, `(C2+C5):C4`, `-C3:C5` I can calculate the constants on the bottom from those on the top. In reverse we have `C1:C1`, `C2:C3`, `C3:-C5`, `C4:(C2+C5)`, `C5:(C2-C3)`. So if I can calculate the constants on the top from those on the bottom and vice versa then these are equally general. That means it's fine to update the test with this new solution.

With all tests passing there I'm happy that this PR works. However what remains is to consider whether the patch here is the best way to go about fixing #4838. For that we should look carefully through `odesimp`, `constant_renumber`, etc and think about the best way to implement this.

XtremeGood commented:
Hmmm, now I get it. I will soon think something till tomorrow to solve this.Thanks!

XtremeGood commented:
Well I get it we gotta implement a system which can understand C1 +C2 = C3 a new constant.

XtremeGood commented:
Well integration constants are not constants but arbitrary constants.Guess we gotta make a class for  that kind of variables.

XtremeGood commented:
Well I asked on the sympy chat and it seems combining coefficients is not implemented in sympy.Although there are functions like odesimp() and constantsimp() but they require arguments to be specified in order to work but I have no way to extract C1,C2.. etc explicitly from the equation yet.But I have a feel that "Wild" might work but I do not understand it fully.

oscarbenjamin commented:
This isn't about combining constants. This PR is a out a bug that happens when renumbering constants at the end of dsolve.

Longer term I think that integration constants shoule be Dummy symbols which would help a bit.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2955,8 +2966,17 @@ def _constant_renumber(expr):
+        exclude = []
+    dummy_ode = Add(*exclude)
+    newconsts = get_numbered_constants(dummy_ode, num=newstartnumber)
+    # FIXME: Why does get_numbered_constants not always return a list?
```
check the code: `return (Cs[0] if num == 1 else tuple(Cs))`. When `num == 1` you get a single constant. So you could use 
```python
if newstartnumber == 1:
    newconsts = [newconsts]
```

smichr replied:
or just use `newconsts = list(iter_numbered_constants(dummy_ode, num=newstartnumber))`.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2861,7 +2872,7 @@ def _conditional_term_factoring(expr):
 
 
-def constant_renumber(expr, symbolname, startnumber, endnumber):
+def constant_renumber(expr, symbolname, startnumber, endnumber, exclude=None):
```
you could use `exclude=[]` which would document what is expected for exclude and then not require the line below to change None into []

oscarbenjamin replied:
Mutable defaults make me suspicious even if it doesn't matter in this context.

smichr commented:
#15992 is in.

oscarbenjamin commented:
I think what's really bothering me here is that the general signature of constant_renumber is wrong. It is inherently assuming that there are sequentially numbered constants in the solution but as the linked issues show we can't make that assumption.

Really constant_renumber should be something like `constant_renumber(sol, constants)` i.e. you pass in a list of which symbols are constants.

smichr commented:
> Really constant_renumber should be something like

That sounds better. I wonder if you could do as you said earlier: use Dummies during the solve process and then rename with non-clashing numbered constants at the end. One would have to watch out for the simplify=False flag which may -- I don't know -- control some aspect of the numbering/renaming.

oscarbenjamin commented:
Looks like it's documented as being internal
[https://github.com/sympy/sympy/blob/85e64019da37177d20bbb0c25550cd86c5797e6a/sympy/solvers/ode.py#L41](https://github.com/sympy/sympy/blob/85e64019da37177d20bbb0c25550cd86c5797e6a/sympy/solvers/ode.py#L41)
so I guess it's fine to change the signature.

oscarbenjamin commented:
It is not sufficient to tell `constant_renumber` which symbols in the solution are constants as there may be other constants in the ODE that don't appear in the solution:
```julia
In [19]: C1 = Symbol('C1')                                                                                                                     

In [20]: eq = f(x).diff(x) * (f(x).diff(x) + C1)                                                                                               

In [21]: eq                                                                                                                                    
Out[21]: 
⎛     d       ⎞ d       
⎜C₁ + ──(f(x))⎟⋅──(f(x))
⎝     dx      ⎠ dx      

In [22]: dsolve(eq)                                                                                                                            
Out[22]: [f(x) = C₁, f(x) = C₁ - C₂⋅x]
```
Here the first solution `f(x)=C1` should be `f(x)=C2` since `C1` is present in the original ODE. That example isn't fully fixed yet by this PR:
```julia
In [4]: dsolve(eq)                                                                                                                             
Out[4]: [f(x) = C₁, f(x) = -C₁⋅x + C₂]
```

Perhaps it should be `constant_renumber(eq, sol)`. The constants in `sol` can be identified from `sol.free_symbols-eq.free_symbols` and can be replaced with those from `get_numbered_constants(eq)`.

Really though when returning multiple solutions like that the constants should be different in each solution. I think we should use Dummy not just internally but in the returned solution.

smichr commented:
So `[f(x) = C₁, f(x) = -C₁⋅x + C₂]` should be `[f(x) = C₂, f(x) = -C₁⋅x + C₂]`

My naive approach would be to just handle this at the outset and let the machinery that is already set up to deal with the non-clashing case continue without modification:

```python
def dsolve(...):
    # watch out for clashing symbols
    conC = lambda eq: [i for i in eq.atoms(Symbol) if 
        i.name.startwith('C') and all(d in string.digits for d in i.name[1:]]
    clashing = conC(eq)
    if clashing:
        reps = dict(zip(clashing, [Dummy() for i in clashing]))
        neweq = eq.xreplace(reps)
        sol = dsolve(neweq, ...)
        newC = conC(sol)
        for c in newC:
            cnew, offset = c, 1
            while cnew in clashing:
                cnew = Symbol(cnew[0] + str(int(cnew[1:]) + offset))
                offset += 1
            reps[cnew] = c
            clashing.append(cnew)
        ureps = dict([tuple(reversed(i)) for i in reps.items()])
        return sol.xreplace(ureps)
```

XtremeGood commented:
Well the integral constants if are strings then they would have "C" in them. If you use sympy.atoms(eq) then you should get {C1,C2....,x,y,z}.If we could filter all elements having C in some way then we can get it. But C1,C2... are not strings.Is there any method to convert them to strings?

XtremeGood commented:
Can we define a function namely "linearly_independent_functions" in sympy.I think that would go the job for many integrals.It should find all the functions in an equation i.e split at plus symbols,and return a set of linearly independendent functions and then we can concatinate/multiply the C's  to return the solution?Can we impleent such a thing?

oscarbenjamin commented:
@smichr On the whole dsolve is already setup fairly well to handle this. I think I can fix it with a simpler patch than the one here (although changing the signature of constant_renumber will mean changing many tests).

@XtremeGood The ode module already does much of that. Bear in mind that what we are working on here needs to apply to nonlinear ODEs as well.

oscarbenjamin committed:
Don't clobber symbols from the ODE in odesimp

oscarbenjamin commented:
I've just pushed an update here that changes the signatures of some internal functions in ode.py to fix the related issues in a cleaner way.

This now also fixes the case mentioned above:
```julia
In [1]: C1 = Symbol('C1')                                                                                                                      

In [2]: eq = f(x).diff(x) * (f(x).diff(x) + C1)                                                                                                

In [3]: eq                                                                                                                                     
Out[3]: 
⎛     d       ⎞ d       
⎜C₁ + ──(f(x))⎟⋅──(f(x))
⎝     dx      ⎠ dx      

In [4]: dsolve(eq)                                                                                                                             
Out[4]: [f(x) = C₂, f(x) = -C₁⋅x + C₂]
```

The changes to signatures are:
```diff
- def odesimp(eq, func, order, constants, hint):
+ def odesimp(ode, eq, func, hint):
```

```diff
- def constant_renumber(expr, symbolname, startnumber, endnumber):
+ def constant_renumber(eq, expr):
```

```diff
- def _handle_Integral(expr, func, order, hint):
+ def _handle_Integral(expr, func, hint):
```

The reason for making these changes is that the signatures of the functions are designed under faulty premises. It is not possible to correctly handle constant renaming by assuming that constants are sequentially numbered (as is the case for `constant_renumber` on master). It is also not sufficient to pass a list of constant symbols to `constant_renumber` since it needs to know not only which symbols in the solution are constants but also which symbols appeared in the ODE so that the newly named constants don't clash with those.

In light of those considerations the simplest reasonable signature is `constant_renumber(eq, sol)` i.e. we pass both the ODE and the solution. Since this is called by `odesimp` we need to pass the ODE into there as well. That makes the `order` and `constants` arguments to `odesimp` redundant so I removed them. The change to `_handle_Integral` is because it doesn't use the `order` argument anyway and having changed the other functions it is no longer convenient to provide that argument.

Todo:
1. Rewrite the docstrings for `odesimp` and `constant_renumber` and check ODE docs.
2. Add test for the case mentioned above.
3. Look for other constant related issues to fix while I'm here (e.g. #11024).

The changes are not backwards compatible if anyone is using these functions but they are documented as being internal. Every call to `constant_renumber` had to be changed in the tests so you can get an idea of the problem looking at the changes there.

Before proceeding with this can anyone confirm that these changes are acceptable for backwards compatibility?

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2899,31 +2915,17 @@ def constant_renumber(expr, symbolname, startnumber, endnumber):
-        )
+        return type(expr)([constant_renumber(eq, e) for e in expr])
+
+    # Symbols in solution bot not ODE are constants
```
bot->but

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2870,8 +2883,11 @@ def _conditional_term_factoring(expr):
 
 
-def constant_renumber(expr, symbolname, startnumber, endnumber):
+def constant_renumber(eq, expr):
```
What about making the signature `(expr, exclude=None)`? The user would then pass `(expr, eq.free_symbols)`.

Is this just an internal function that we are free to change?

oscarbenjamin replied:
The docstring for the ode module says that these are internal functions:
[https://github.com/sympy/sympy/blob/2d4eb2fd9f8333bb1992c31069f256db542b589b/sympy/solvers/ode.py#L30](https://github.com/sympy/sympy/blob/2d4eb2fd9f8333bb1992c31069f256db542b589b/sympy/solvers/ode.py#L30)

We could do have `exclude=None` but the signature would need to be:
```python
def constant_renumber(expr, constants, exclude):
```
so that we know which symbols in expr are constants *and* which symbols not in expr would still clash with the ODE.

smichr replied:
> docstring for the ode module says that these are internal functions

Then I think we are ok changing them. The approach here seems much less fragile and more user friendly.

I'm not sure I understand the subtlety of the distinction between `exclude` and `foo.free_symbols`. It seems to me that `eq` is included with `expr` (in `(eq, expr)`) so you can identify which constants were introduced into `expr` that were not in `eq`...I am just suggesting that you let the user tell you what symbols to ignore with `exclude` instead of passing `eq`.

oscarbenjamin replied:
There is no subtlety in the distinction. At the moment this PR allows you to do:
```python
sol_renumbered = constant_renumber(eq_ode, sol)
```
If we changed it to have `(expr, constants, exclude)` as arguments then the equivalent would be:
```python
sol_renumbered = constant_renumber(sol,
     constants=sol.free_symbols - eq_ode.free_symbols,
     exclude=eq_ode.free_symbols)
```
That might seem cumbersome for the caller but this function is called twice in the whole module (including once recursively) so it's not a big deal for the caller to prepare these arguments. The tests call this function a lot but we could make a helper there.

So how about:
```python
def constant_renumber(expr, constants, exclude=None):
```
where exclude is an optional list (not needed in the tests) and constants is a list of symbols.

Then in the tests we can do:
```diff
-    sol8s = constant_renumber(sol8, 'C', 1, 4)
+    sol8s = constant_renumber(sol8, symbols('C1:5'))
```

The change to `odesimp` can then just be a new optional `exclude` argument.

oscarbenjamin replied:
Another possibility (bearing in mind #11024) is that the signature can be:
```python
def constant_renumber(expr, old_constants, new_constants_iter=None):
```
Then inside `dsolve` we can pass `new_constants_iter=iter_numbered_symbols(eq_ode, start=1, prefix='C')` which will generate new constants not found in the ODE. We can then document that users calling `dsolve` with their own constants iterator will need to ensure that it only generates symbols not found in the ODE (e.g. by using `iter_numbered_symbols`).

Yet another possibility would be
```python
def constant_renumber(expr, old_constants, exclude=None, new_constants_iter=None):
```
and `constant_renumber` can do the checking to ensure that constants coming from `new_constants_iter` don't clash with those in `exclude`. That would be slightly friendlier on `dsolve` users who want to pass in their own constant iterator.

smichr replied:
I would make it
```python
def constant_renumber(sol, exclude=None, cons=None):
    exclude = set(exclude or [])
    constants=sol.free_symbols - set(exclude or [])
    ...
```
And call it as
```python
constant_renumber(sol, eq_ode.free_symbols, numbered_constants('C'))
```
> need to ensure that it only generates symbols not found in the ODE

I would just keep generating symbols until no clashes are found. So the user basically decides on what symbol name will be used and numbered symbols are supplied that are not already in use.

oscarbenjamin replied:
> I would make it
> 
> ```python
> def constant_renumber(sol, exclude=None, cons=None):
> ```

Yeah, that looks good. Maybe exclude should be called `nonconstants` or something.

smichr replied:
> should be called `nonconstants` or something.

or `variables`

Will `constant_renumber` only target numbered constants or would it use numbered constants for any constants that weren't excluded? e.g. 
```python
>> constant_renumber(x + y, exclude=y, numbered_symbols('c'))
c0 + y
```

oscarbenjamin replied:
With the signature that you have just proposed any symbol found in sol but not in exclude will be replaced with a symbol from cons (which will default to C1, C2, ...) so that would be:
```python
>>> constant_renumber(x + y, exclude=[y], iter_numbered_symbols(eq_ode, 'c'))
c1 + y
```

I think `variables` here is not quite right. In `Eq(f(x).diff(x), a*x)` we would normally refer to `f` and `x` as the dependent and independent variables. The symbol `a` is a constant but not an integration constant. We need `exclude` to be the list of all symbols in the solution that are not integration constants.

We're probably overthinking this though since this is an internal function. I will go make the changes.

smichr replied:
I guess I am thinking of something more general: a function that replaces constants with numbered symbols that don't clash with any variables in the original expression. That is why I use `numered_constants('c')` rather than refer to `eq_ode` in `iter_numered_symbols`. So the function basically needs to know the original expression and which symbols *not* to target and then just numbers everything else, making sure not to clash with an existing symbol.
```python
def syms_numbered(current, fixed, syms):
    old = list(ordered(set(current) - set(fixed)))
    new = []
    for i in old:
        while True:
            s = next(syms)
            if s not in fixed:
                new.append(s)
                break
    return dict(zip(old, new))  # use as eq.xreplace(rv)
```
e.g.
```python
>>> syms_numbered([x,y,var('c1')],[y],numbered_symbols('x'))
{c1: x0, x: x1}
```

oscarbenjamin replied:
That's more or less where I ended up yesterday. I didn't push at the time because of the build queue. I'll push it now.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2318,9 +2320,20 @@ def _expand(expr):
+    # Don't clobber the symbols already in the ODE. This is important when the
+    # input ODE derives from output of a previous call to dsolve (issue #4838)
+    if iterable(eq):
+        symbols_eq = set()
```
this is not currently used (though I kind of like this idea of using `exclude`)

oscarbenjamin replied:
I guess I missed "general cleanup" off the TODO list. I want to get agreement on the signatures first though.

oscarbenjamin replied:
That's left over from the previous version of this PR.

oscarbenjamin committed:
Rejig constant_renumber signature

smichr committed:
Merge branch 'master' into dsolve_constantsimp

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2883,7 +2883,7 @@ def _conditional_term_factoring(expr):
 
 
-def constant_renumber(eq, expr):
+def constant_renumber(expr, nonconstants=None, newconstants=None):
```
nonconstants -> variables?

oscarbenjamin replied:
odesymbols?

smichr replied:
I prefer a generic name since this function could be used by anyone; it has use beyond ODE.

oscarbenjamin replied:
Can you give an example of somewhere else it could be used? That might help with a name. Variables doesn't seem right in this particular context.

oscarbenjamin replied:
Also if it has more general use it should probably be moved somewhere else.

smichr replied:
> Can you give an example of somewhere else it could be used?

If you have `f(x, a, b, c) = a*(b + c*x)` and you want to model that as a function of 'x' with the fewest symbols possible you would want to collapse constants together like `constantsimp` does and the process would have nothing to do with odes. The model could be `c1 + c2*x` in this particular case with `c1 = a*b` and `c2 = a*c`.

smichr replied:
cf #4435

oscarbenjamin replied:
That makes sense but the frontend for that should be something that also calls constantsimp. For now these inputs to the function are sufficient to correctly do its job for dsolve.

oscarbenjamin commented:
Looks like I forgot to update the tests in test_constantsimp.

oscarbenjamin committed:
Fix test_constantsimp for new constant_renumber

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2899,31 +2915,30 @@ def constant_renumber(expr, symbolname, startnumber, endnumber):
+    else:
+        nonconstants = set()
+        # Any Cn is a constant...
+        isconstant = lambda s: s.startswith('C') and s[1:].isdigit()
```
hmmm...I didn't know `isdigit` would handle more than one character ('123'.isdigit() -> True)

smichr commented:
TODO added to OP

asmeurer commented:
> Longer term I think that integration constants shoule be Dummy symbols which would help a bit.

Dummy symbols are harder to access for an end user. Maybe that's less important now that dsolve natively supports initial conditions. Also even though `Dummy("C1") != Dummy("C1")`, they would still print the same and it could be confusing to end-users. 

However, there is some precedent now for functions returning results with dummy symbols (usually as bound variables), so it could be considered. For instance, taking the element of a MatMul of symbolic shape produces a Sum with Dummy index.

oscarbenjamin committed:
Cleanup after rejigging constant_renumber

oscarbenjamin commented:
When I say that they should be dummy symbols I really mean that they should be a special kind of integration constant symbol that compares unequal to others like Dummy symbols do. There should be a function for extracting the constants from an expression so that you can do:

```julia
In [1]: sol = dsolve(f(x).diff(x, 2))                                                                                             

In [2]: sol                                                                                                                       
Out[2]: f(x) = C₁ + C₂⋅x

In [3]: get_integration_constants(sol)
Out[3]: [C₁, C₂]
```
(or something like that)

This would require all the ODE tests to be changed because you couldn't just do
```python
assert dsolve(f(x).diff(x)) == Eq(f(x), C1)
```

Otherwise I don't see what need you could possibly have for the integration constants to compare equal. If they don't come from the same integration then comparing equal is an error:
```julia
In [8]: sol1 = dsolve(f(x).diff(x))                                                                                               

In [9]: sol2 = dsolve(f(x).diff(x)-f(x))                                                                                          

In [10]: sol1                                                                                                                     
Out[10]: f(x) = C₁

In [11]: sol2                                                                                                                     
Out[11]: 
           x
f(x) = C₁⋅ℯ 

In [13]: sol1.rhs / sol2.rhs                                                                                                      
Out[13]: 
 -x
ℯ  
```

oscarbenjamin commented:
@smichr I've pushed a commit that addresses the TODO list above.

I also put "add tests" in the TODO list myself but I now see that I already added the tests from the relevant issues.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2964,13 +2984,16 @@ def _constant_renumber(expr):
     expr = _constant_renumber(expr)
+
+    # Don't renumber symbols present in the ODE.
+    constants_found = [c for c in constants_found if c not in variables]
```
You can use `list(filter_symbols(constants_found, variables))` -- avail from iterables.

oscarbenjamin replied:
I don't think that's an improvement. I can see why `filter_symbols` exists (for the infinite iterator situation) but for the finite case a list comprehension is clearer.

oscarbenjamin replied:
Okay I guess the list comprehension would be quadratic if variables was a list rather than a set. It always a set as called currently but someone might call it with a list (as shown in the docstring) so I should either use `filter_symbols` or ensure that it is a set.

oscarbenjamin replied:
Since I was already calling `set(variables)` above I just changed it to `variables = set(variables)` so that it will always be a set.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2870,16 +2870,23 @@ def _conditional_term_factoring(expr):
-    Renumber arbitrary constants in ``expr`` to have numbers 1 through `N`
-    where `N` is ``endnumber - startnumber + 1`` at most.
-    In the process, this reorders expression terms in a standard way.
+    FIXME: Need to redo this docstring. This now renumbers constants in expr
```
FIXME still here. Also the bot->but typo needs fixing below

oscarbenjamin commented:
Thanks for the review @smichr I will fix up the docstring. I think I just spotted a problem with this that will need a bit of testing and possibly a fix.

oscarbenjamin committed:
Ensure that variables is set in constant_renumber

oscarbenjamin commented:
I've just pushed an update. The issue I thought I had spotted turned out to be nothing. I just confused myself that there was a problem :)

I've fixed the docstring. I also changed the global variable to be nonlocal while I was there.

smichr commented on sympy/solvers/ode.py:
```diff
@@ -2955,7 +2954,7 @@ def _constant_renumber(expr):
 
         """
-        global newstartnumber
+        nonlocal newstartnumber
```
```
Compiling sympy/solvers/ode.py ...
  File "sympy/solvers/ode.py", line 2957
    nonlocal newstartnumber
                          ^
SyntaxError: invalid syntax
```

smichr replied:
from [here](https://stackoverflow.com/questions/14264313/syntax-error-on-nonlocal-statement-in-python): nonlocal only works in Python 3; it is a new addition to the language.

oscarbenjamin replied:
Sorry, I really should have checked that before pushing. I've just pushed again

oscarbenjamin committed:
Note to remove global statement

smichr commented:
+1

oscarbenjamin commented:
Thanks @smichr !
